gobblin-runtime/src/main/java/gobblin/runtime/api/JobExecutionState.java

```diff
@@ -36,4 +54,86 @@ public JobExecutionState(JobSpec jobSpec, JobExecution jobExecution) {
     return Collections.unmodifiableMap(this.executionMetadata);
   }
 
+  @Override
+  public String toString() {
+    return Objects.toStringHelper(JobExecutionState.class.getSimpleName())
+            .add("jobExecution", this.jobExecution)
+            .add("runningState", this.runningState)
+            .add("stage", this.stage)
+            .toString();
+  }
+
+  public void setRunningState(JobState.RunningState runningState) {
+    switch (runningState){
+      case PENDING: this.switchToPending(); break;
+      case RUNNING: this.switchToRunning(); break;
+      case SUCCESSFUL: this.switchToSuccessful(); break;
+      case FAILED: this.switchToFailed(); break;
+      case COMMITTED: this.switchToCommitted(); break;
+      case CANCELLED: this.switchToCancelled(); break;
+      default: throw new Error("Unexpected state");
+    }
+  }
+
+  public void switchToPending() {
+    Preconditions.checkState(null == this.runningState, "unexpected state:" + this.runningState);
```

Comment:
nit: `String.format("unexpected stage transition: %s to %s", this.runningState, PENDING)`, also for the other similar methods.